kyo-prelude/shared/src/main/scala/kyo2/kernel/Runtime.scala

```diff
@@ -0,0 +1,121 @@
+package kyo2.kernel
+
+import internal.*
+import java.util.Arrays
+import kyo2.isNull
+import scala.annotation.tailrec
+import scala.util.control.NonFatal
+
+final class Runtime(initDepth: Int, initState: Runtime.State):
+    import Runtime.State
+
+    private val owner         = Thread.currentThread()
+    private[kernel] var depth = initDepth
+    private var traceIdx      = initState.traceIdx
+    private val trace         = Runtime.copyTrace(initState.trace, traceIdx)
+
+    private def enter(_frame: Frame): Int =
+        if (Thread.currentThread eq owner) && depth < maxStackDepth then
+            pushFrame(_frame)
+            val depth = this.depth
+            this.depth = depth + 1
+            depth + 1
+        else
+            -1
+        end if
+    end enter
+
+    private def pushFrame(frame: Frame): Unit =
+        val traceIdx = this.traceIdx
+        trace(traceIdx & (maxTraceFrames - 1)) = frame
+        this.traceIdx = traceIdx + 1
+    end pushFrame
+
+    private[kernel] def clearTrace(): Unit =
+        traceIdx = 0
+
+    private def exit(depth: Int): Unit =
+        this.depth = depth - 1
+
+    private[kernel] def save(values: Values) =
+        State(Runtime.copyTrace(trace, traceIdx), traceIdx, values)
+end Runtime
+
+object Runtime:
+
+    implicit def get: Runtime = local.get()
```

Comment:
Assuming this should not be called often (if ever) by users, should it be `private[kyo]`?

Reply:
It needs to be available in user code to resolve the `Safepoint` (previously `Runtime`) when performing transformations.